### PR TITLE
Content info module

### PR DESCRIPTION
**Pull request by dhineshwiz**

### CODE REVIEW

File: lib/content-info/index.js [hunk 1]
```diff
@@ -0,0 +1,375 @@
...
+     * @type {String}
+     */
+    BINARY = 'binary',
+
+    /**
+     * @private
+     * @const
+     * @type {String}
+     */
+    UTF8 = 'utf8',
```
Comment (kamalaknn) [inline lib/content-info/index.js]: CHARSET_UTF8

File: lib/content-info/index.js [hunk 2]
```diff
@@ -0,0 +1,375 @@
...
+
+    /**
+     * @private
+     * @const
+     * @type {String}
+     */
+    TEXT_PLAIN = 'text/plain',
```
Comment (kamalaknn) [inline lib/content-info/index.js]: CONTENT_TYPE_TEXT_PLAIN

File: lib/content-info/index.js [hunk 3]
```diff
@@ -0,0 +1,375 @@
...
+
+    /**
+     * Enum for all the Content Headers
+     * @private
+     * @const
+     * @enum {String} CONTENT_HEADERS
+     */
+    CONTENT_HEADERS = {
```
Comment (kamalaknn) [inline lib/content-info/index.js]: HEADERS

File: lib/collection/response.js [hunk 4]
```diff
@@ -316,8 +327,7 @@ _.assign(Response.prototype, /** @lends Response.prototype */ {
      *   _accuratelyDetected: !normalised.orphan // this being true implies worse case (raw render)
      *   detected: {} // same as root object, but based on what is detected from content
      * }
-     *
-     * @todo write unit tests
+     * @deprecated Use  {@link Response#contentInfo} contentInfo instead. To be removed in 4.0
```
Comment (kamalaknn) [inline lib/collection/response.js]: Can you flip the sentences? Looks like `contentInfo` will be removed in `4.0`.

File: lib/collection/response.js [hunk 5]
```diff
@@ -502,7 +512,7 @@ _.assign(Response, /** @lends Response */ {
      * @param {String|Header} disposition
      * @returns {Object}
      *
-     * @todo write unit tests
+     * @deprecated Use  {@link Response#contentInfo} contentInfo instead. To be removed in 4.0
```
Comment (kamalaknn) [inline lib/collection/response.js]: Same as above

File: lib/content-info/index.js [hunk 6]
```diff
@@ -0,0 +1,375 @@
...
+            charset: normalized.charset || UTF8,
+            extension: mimeType.extension(normalized.source) || E
+        };
+    },
+
+    /**
+     * Parses Content disposition header, and returns file name and extension
+     *
+     * @private
+     * @param {String} dispositionHeader - Content-disposition Header from the response
```
Comment (kamalaknn) [inline lib/content-info/index.js]: Nullable

File: lib/content-info/index.js [hunk 7]
```diff
@@ -0,0 +1,375 @@
...
+        return characterDecoders[charset](binary);
+    },
+
+    /**
+     * Takes the content-type header value and performs the mime sniffing with known mime types.
+     * If content-type header is not present, detects the mime type from the response stream or response body
+     * If content-type is not provided and not able to detect, then text/plain is taken as default
+     *
+     * @private
+     * @param {String} contentType - The value of content type header
```
Comment (kamalaknn) [inline lib/content-info/index.js]: Nullable

File: lib/content-info/index.js [hunk 8]
```diff
@@ -0,0 +1,375 @@
...
+     * @param {Stream|String} response - The response stream or body, for which content-info should be determined
```
Comment (kamalaknn) [inline lib/content-info/index.js]: Nullable

File: lib/content-info/index.js [hunk 9]
```diff
@@ -0,0 +1,375 @@
...
+
+
+module.exports = {
+
+    /**
+     * Extracts content related information from response.
+     * Includes response mime information, character set and file name.
+     *
+     * @param {Response} response - response instance
+     * @private
```
Comment (kamalaknn) [inline lib/content-info/index.js]: Move up

File: test/unit/content-info.test.js [hunk 10]
```diff
@@ -0,0 +1,465 @@
...
+        it('should not get ReDos by fileNameRegex for long patterns of ASCII and non-ASCII chars unequal split parse', function () {
+            this.timeout(1000);
+            var filenameHeader = 'attachment;filename="' + 'helloooo你好'.repeat(1e5) + '"',
+                headerPart = contentInfo.regexes.fileNameRegex.exec(filenameHeader);
+            expect(headerPart).eql(null);
+        });
+
+        // eslint-disable-next-line max-len
+        it('should not get ReDos by fileNameRegex for long patterns of ASCII and non-ASCII chars unequal split multiple', function () {
+            this.timeout(1000);
```
Comment (kamalaknn) [inline test/unit/content-info.test.js]: can you move this to the `describe`?

File: lib/collection/response.js [hunk 11]
```diff
@@ -292,6 +293,16 @@ _.assign(Response.prototype, /** @lends Response.prototype */ {
         return LJSON.parse(this.text(), reviver, strict);
     },
 
+    /**
+     * Extracts mime type, format, charset, extension and filename of the response content
+     * A fallback of default filename is given, if filename is not present in header
```
Comment (kamalaknn) [inline lib/collection/response.js]: not present in `content-disposition` header

File: lib/content-info/index.js (hunk 8 above)
Comment (dhineshwiz) [reply lib/content-info/index.js]: I think response can't be nullable

File: CHANGELOG.md [hunk 12]
```diff
@@ -1,5 +1,9 @@
 # Postman Collection SDK Changelog
 
+### unreleased (Jun 1, 2018)
```
Comment (kamalaknn) [inline CHANGELOG.md]: Unreleased should not have date

File: lib/content-info/index.js [hunk 13]
```diff
@@ -0,0 +1,375 @@
...
+        if (!encodedFileName) {
+            return;
+        }
+
+        if (!characterDecoders[charset]) {
+            return;
+        }
+
+        // Decodes the hexadecimal numbers to charCode
+        var binary = encodedFileName.replace(regexes.hexCharMatchRegex, decodeHexcode);
```
Comment (shamasis) [inline lib/content-info/index.js]: Do we need this interstitial variable?

File: test/unit/content-info.test.js [hunk 14]
```diff
@@ -0,0 +1,629 @@
...
+            .eql({
+                charset: 'utf8',
+                fileExtension: 'json',
+                mimeFormat: 'json',
+                mimeType: 'text',
+                fileName: 'testResponse.json'
+            });
+    });
+
+    describe('edge cases', function () {
```
Comment (kamalaknn) [inline test/unit/content-info.test.js]: Drop the `edge cases` describe block and inline these tests wherever applicable,

Comment (shamasis) [reply test/unit/content-info.test.js]: Lol. This is hilarious. Can we drop the “edge case” please.

Comment (shamasis) [reply test/unit/content-info.test.js]: Lol. This is hilarious. Can we drop the “edge case” please.

File: test/unit/content-info.test.js [hunk 15]
```diff
@@ -0,0 +1,629 @@
...
+        it('Should take default filename if content-disposition header value is empty', function () {
```
Comment (shamasis) [inline test/unit/content-info.test.js]: Should we also not have test cases where one or both of the headers are missing?
